StabilityPackage.wl

```diff
@@ -193,36 +276,168 @@ Return[as]
+kblow=OptionValue["krange"][[1]];
+kbhigh=OptionValue["krange"][[2]];
+kgrid=Join[fSpace[ktarget*kblow,ktarget*kbhigh,nstep],-fSpace[ktarget*kblow,ktarget*kbhigh,nstep/2]];
 Return[kgrid];
 ];
 
+
+
 (*Run buildkGrid and SCalcScale for several radial bins.*)
-kAdapt[infile_,rstr_,rend_,testE_,hi_,nstep_]:= Module[{kl,evout,data,singleRadiusData,ea,kvar,evals,pot,ktarget,S},
+Options[kAdapt]={"dh"-> True,"xflavor"-> False,"ktarget"-> 0.,"krange"-> {10.^-3,10.}};
+kAdapt[infile_,rstr_,rend_,testE_,hi_,nstep_,OptionsPattern[]]:= Block[{kl,evout,data,singleRadiusData,ea,kvar,evals,pot,S},
 data=ImportData[infile];
 evout=
 Reap[
 	Do[
 		singleRadiusData = SelectSingleRadius[data,rx];
-		ea=getEquations[singleRadiusData,testE,hi,kvar];
-		S=stabilityMatrix[data,ea];
-		ktarget=(S/.kvar->0)[[1,1]];
-		kl=buildkGrid[ktarget,nstep];
+		ea=getEquations[singleRadiusData,testE,hi,kvar,"dh"-> OptionValue["dh"],"xflavor"-> OptionValue["xflavor"]];
+		S=stabilityMatrix[data,ea,"xflavor"-> OptionValue["xflavor"]];
+		kl=buildkGrid[singleRadiusData,nstep,"ktarget"-> OptionValue["ktarget"],"krange"-> OptionValue["krange"],"xflavor"-> OptionValue["xflavor"]];
 		Do[
 			evals=Sort[Im[evscale[kl[[kx]],S,kvar]],Greater];
-			pot=potential[data,ea]/.kvar->kl[[kx]];
+			pot=siPotential[singleRadiusData,"xflavor"-> OptionValue["xflavor"]];
 			Sow[{data["radius"][[rx]],kl[[kx]],evals[[1]],pot}]; 
 		,{kx,1,Length[kl]}] (*close do over ktargets*)
 	,{rx,rstr,rend}] (*close do over r*)
 ][[2,1]];
 Return[evout] (*Close reap over r*)
 ]; (*close module*)
 
+
+
+
+
+GDValue[tm_]:=Block[{cs,rrow,rcol,drow,dcol,mv,reg,\[Epsilon],tms},
+\[Epsilon]=$MachineEpsilon;
+tms=Expand[(tm/\[Epsilon])];
+cs=Diagonal[tms];
+rrow=Total[Abs[tms],{2}]-Abs[cs];
+rcol=Total[Abs[tms],{1}]-Abs[cs];
+drow=MapThread[Disk,{ReIm[cs],rrow}];
+dcol=MapThread[Disk,{ReIm[cs],rcol}];
+reg=RegionIntersection[Apply[RegionUnion,drow],Apply[RegionUnion,dcol]];
+mv=\[Epsilon] Max[MeshCoordinates[BoundaryDiscretizeRegion[reg,{{-1,1},{0,1}}]][[All,2]]];
+Return[mv]];
+
+
+GDdata[infile_,rstr_,rend_,testE_,hi_,nstep_]:= Module[{kl,evout,data,singleRadiusData,ea,kvar,evals,pot,ktarget,S,gdout,out,ea2,S2},
+data=ImportData[infile];
+out=
+Reap[
+	Do[
+		singleRadiusData = SelectSingleRadius[data,rx];
+		ea=getEquations[singleRadiusData,testE,hi,kvar];
+		S=stabilityMatrix[data,ea];
+		ktarget=(S/.kvar->0)[[1,1]];
+		kl=buildkGrid[ktarget,nstep];
+		Do[
+		ea2=getEquations[singleRadiusData,testE,hi,kx];
+		S2=stabilityMatrix[data,ea];
+			gdout=GDValue[S2]; (*this isn't updating k correctly*)
+			Sow[{data["radius"][[rx]],kl[[kx]],gdout}]; 
+		,{kx,1,Length[kl]}] (*close do over ktargets*)
+	,{rx,rstr,rend}] (*close do over r*)
+][[2,1]];
+Return[out] (*Close reap over r*)
+		
+]
+
+eBoxFitSingleRadius[data_,ri_,species_,guesses_]:=Module[{ebox,es1box,es2box,es3box,ne,Fe,Pre,ane,aFe,aPre,datasr,br,ei,aei,ag,aag,\[Beta]g,a\[Beta]g,\[Chi]g,a\[Chi]g,foc1234,afoc1234,g0,ag0},
```

Comment:
I think it would be better to pass this function a dataSingleRadius, rather than the full data object and a radial index.